File: mq/pub.go
package mq

const Sub_ChainCfg = "ChainCgf"
const Sub_ContractAbi = "ContractCfg"
const Sub_ContractRule = "ContractRule"
const Sub_RiskRule = "RiskRule"
const Sub_ChainTx = "ChainTx"

const Sub_RiskRuleReply = "RiskRuleReply"
const (
	OptAdd    = "add"
	OptUpdate = "update"
	OptDelete = "delete"
	OptCheck  = "check"
)

// Sub_ChainTx
type ChainTxMsg struct {
	ChainId     int64  `json:"chainId"`
	TxHash      string `json:"txHash"`
	TxIdx       int    `json:"txIdx"`
	BlockNumber int    `json:"blockNumber"`
	BlockHash   string `json:"blockHash"`
	LogIdx      int    `json:"logIdx"`

	From     string `json:"from"`
	To       string `json:"to"`
	Contract string `json:"contract"`
	Value    string `json:"value"`
}

// /Sub_ChainCfg
type ChainCfgReq struct {
	Opt     string `json:"opt"`
	Id      int64  `json:"id"`
	ChainId uint64 `json:"chainId"`
	Coin    string `json:"coin"`
	Rpc     string `json:"rpc"`
}

// ContractAbi
type ContractAbiReq struct {
	Opt             string `json:"opt"`
	Id              int64  `json:"id"`
	ContractAddress string `json:"contractAddress"`
	ChainId         string `json:"chainId"`
}

func (s *ContractAbiReq) IsValid() bool {
	if s.Opt == "" || s.Id <= 0 || s.ContractAddress == "" || s.ChainId == "" {
		return false
	}
	return true
}

// /ContractRule
type ContractRuleReq struct {
	// 'add' | 'update' | 'delete'
	Opt             string `json:"opt"`
	Id              int64  `json:"id"`
	ContractAddress string `json:"contractAddress"`
	ChainId         string `json:"chainId"`
}

func (s *ContractRuleReq) IsValid() bool {
	if s.Opt == "" || s.Id <= 0 || s.ContractAddress == "" || s.ChainId == "" {
		return false
	}
	return true
}

// /RiskRule
type RiskCtrlRuleReq struct {
	//up/del/verify
	Opt string `json:"opt"`
	// Salience int    `json:"salience"`
	// RuleName string `json:"ruleName"`
	// RuleStr  string `json:"ruleStr"`
	// Desc     string `json:"desc"`
	Id       int  `json:"id"`
	IsEnable bool `json:"isEnable"`
}

func (s *RiskCtrlRuleReq) IsValid() bool {
	// if s.Opt == "" || s.RuleStr == "" || s.ChainId == "" {
	// 	return false
	// }
	return true
}

// /RiskRuleReply
type RiskRuleReplyReq struct {
	Opt      string `json:"opt"`
	RuleName string `json:"ruleName"`
	RuleStr  string `json:"ruleStr"`
}
type RiskRuleReplyRsp struct {
	Code int    `json:"code"`
	Msg  string `json:"message"`
}

File: mq/pub.ts

const Sub_ChainCfg = 'ChainCfg';
const Sub_ContractAbi = 'ContractCfg';
const Sub_ContractRule = 'ContractRule';
const Sub_RiskRule = 'RiskRule';

const Sub_RiskRuleReply = 'RiskRuleReply';
///
var OptAdd = 'add';
var OptUpdate = 'update';
var OptDelete = 'delete';
var OptCheck = 'check';
export {
  Sub_ChainCfg,
  Sub_ContractAbi,
  Sub_ContractRule,
  Sub_RiskRule,
  Sub_RiskRuleReply,
  OptAdd,
  OptUpdate,
  OptDelete,
  OptCheck,
};

///
export interface Msg {
  sub: string;
  opt: string;
  version: number;
  data: any;
  jsonPatch: string;
}
function buildMsg(sub:string, type:string, version :  number, data: any) : Msg {
  return {
    sub: sub,
    opt: type,
    version: version,
    data: data,
    jsonPatch: JSON.stringify(data),
  };
}
let chainCfgVersion = 0

export function buildChainCfg(type: string, data: any): Msg {
  // let msg: Msg = {
  //   sub: Sub_ChainCfg,
  //   opt: type,
  //   version: chainCfgVersion,
  //   data: data,
  //   jsonPatch: JSON.stringify(data),
  // };
  let msg = buildMsg(Sub_ChainCfg, type, chainCfgVersion, data)
  chainCfgVersion = chainCfgVersion + 1;
  if (chainCfgVersion > 100000000) {
    chainCfgVersion = 0;
  }
  return msg;
}

// ContractAbi
// export interface ContractAbiReq {
//   sub: string;
//   opt: string;
//   id: number;
//   contractAddress: string;
//   chainId: string;
// }
let contractAbiVersion = 0;
export function buildContractAbi(type: string, data: any): Msg {
  // let msg: Msg = {
  //   sub: Sub_ContractAbi,
  //   opt: type,
  //   data: data,
  //   // id: id,
  //   // contractAddress: contractAddress,
  //   // chainId: chainId,
  // };
  let msg = buildMsg(Sub_ContractAbi, type, contractAbiVersion, data)
  contractAbiVersion = contractAbiVersion + 1;
  if (contractAbiVersion > 100000000) {
    contractAbiVersion = 0;
  }
  return msg;
}
// export function isValidContractAbiReq(s: Msg): boolean {
//   if (
//     s.opt === '' ||
//     s.id <= 0 ||
//     s.contractAddress === '' ||
//     s.chainId === ''
//   ) {
//     return false;
//   }
//   return true;
// }

// /ContractRule
// export interface ContractRuleReq {
//   sub: string;
//   // 'add' | 'update' | 'delete'
//   opt: string;
//   id: number;
//   contractAddress: string;
//   chainId: string;
// }

// export function buildContractRule(
//   type: string,
//   data: any
//   // id: number,
//   // chainId: string,
//   // contractAddress: string
// ): Msg {
//   // let msg: Msg = {
//   //   sub: Sub_ContractRule,
//   //   opt: type,
//   //   data: data,
//   //   // chainId: chainId,
//   //   // contractAddress: contractAddress,
//   //   // id: id,
//   // };
//   let msg = buildMsg(Sub_ContractAbi, type, data, contractAbiVersion)
//   contractAbiVersion = contractAbiVersion + 1;
//   if (contractAbiVersion > 100000000) {
//     contractAbiVersion = 0;
//   }
//   return msg;
// }
// export function isValidContractRuleReq(s: ContractRuleReq): boolean {
//   if (
//     s.opt === '' ||
//     s.id <= 0 ||
//     s.contractAddress === '' ||
//     s.chainId === ''
//   ) {
//     return false;
//   }
//   return true;
// }

// /RiskRule
// export interface RiskCtrlRuleReq {
//   sub: string;
//   //up/del/verify
//   opt: string;
//   id: number;
//   isEnable: boolean;
// }
let riskCtrlRuleVersion = 0;
export function buildRiskCtrlRule(
  type: string,
  data: any
): Msg {
  let msg = buildMsg(Sub_RiskRule, type, riskCtrlRuleVersion, data)
  riskCtrlRuleVersion = riskCtrlRuleVersion + 1;
  if (riskCtrlRuleVersion > 100000000) {
    riskCtrlRuleVersion = 0;
  }
  return msg;
}
// export function isValidRiskCtrlRuleReq(s: RiskCtrlRuleReq): boolean {
//   // if (s.opt === '' || s.ruleStr === '' || s.sceneNo === '' || s.id <= 0) {
//   //   return false;
//   // }
//   return true;
// }

// /RiskRuleReply
export interface RiskRuleReplyReq {
  sub: string;
  opt: string;
  chainId: string;
  ruleName: string;
  ruleStr: string;
}
export function buildRiskRuleReply(
  chainId: string,
  ruleStr: string,
  ruleName: string
): RiskRuleReplyReq {
  let data: RiskRuleReplyReq = {
    sub: Sub_RiskRuleReply,
    opt: OptCheck,
    chainId: chainId,
    ruleStr: ruleStr,
    ruleName: ruleName,
  };
  return data;
}
export interface RiskRuleReplyRes {
  Code: number;
  Msg: string;
}
///
///
///
